src/dupe_finder.cc

```diff
@@ -0,0 +1,242 @@
+  char  result[4] = { ' ', ' ', ' ', ' ' };
+  for (uint8_t dest_offset = 0; dest_offset < 4 && *str; ++dest_offset) {
+    uint8_t ch = *str++;
+    if (ch == '%') {
+      ch = value_for_hex_char(*str++) << 4;
+      ch |= value_for_hex_char(*str++);
+    }
+    result[dest_offset] = ch;
+  }
+  
+  be_uint32_t  r;
+  memcpy(&r, result, 4);
+  return r;
+}
+
+int main(int argc, const char** argv) {
+  try {
+    if (argc < 2) {
+      print_usage();
+      return 2;
+    }
+    
+    // Process command line args
+    vector<const char*>     input_filenames;
+    set<uint32_t>           input_res_types;
+    bool                    use_data_fork = false;
+    bool                    delete_duplicates = false;
+    bool                    make_backup = false;
+    
+    for (int x = 1; x < argc; x++) {
+      if (!strncmp(argv[x], "--", 2)) {
+        if (!strcmp(argv[x], "--data-fork")) {
+          use_data_fork = true;
+        } else if (!strcmp(argv[x], "--delete")) {
+          delete_duplicates = true;
+        } else if (!strcmp(argv[x], "--backup")) {
+          make_backup = true;
+        } else if (!strncmp(argv[x], "--target-type=", 14)) {
+          input_res_types.insert(parse_cli_type(&argv[x][14]));
+        } else {
+          throw invalid_argument(string_printf("unknown option: %s", argv[x]));
+        }
+      } else {
+        input_filenames.push_back(argv[x]);
+      }
+    }
+
+    if (input_filenames.empty()) {
+      print_usage();
+      return 2;
+    }
+    
+    // Load resource files
+    vector<InputFile>  input_files;
+    for (const char* basename : input_filenames) {
+      string filename = basename;
+      if (!use_data_fork) {
+        filename += "/..namedfork/rsrc";
+      }
+      auto    file_info = stat(filename);
+      if (isfile(file_info) && file_info.st_size > 0) {
+        input_files.push_back({ basename, parse_resource_fork(load_file(filename)), 0 });
+      } else {
+        fprintf(stderr, "Input file %s does not exist, is empty or is not a file\n", filename.c_str());
+      }
+    }
+    
+    // Gather existing resource types
+    if (input_res_types.empty()) {
+      for (const InputFile& file : input_files) {
+        for (uint32_t type : file.resources.all_resource_types()) {
+          input_res_types.insert(type);
+        }
+      }
+    }
+    
+    // Find duplicates, one resource type at a time
+    uint32_t  num_duplicates = 0;
+    for (uint32_t res_type : input_res_types) {
+      // Hash all resources of type `res_type` and group them by hash
+      unordered_map<size_t, vector<Resource>> hashed_resources;
+      for (InputFile& file : input_files) {
+        for (int16_t res_id : file.resources.all_resources_of_type(res_type)) {
+          auto    resource = file.resources.get_resource(res_type, res_id);
+          size_t  hash = std::hash<string>()(resource->data);
+          hashed_resources[hash].push_back(Resource{ file, resource, false });
+        }
+      }
```

Comment:
Then the unordered_map eats the duplicates because it stores only one value per key. Your proposal works with an unordered_multimap. Then we can iterate over the groups with `equal_range` for output/deletion. I originally had a similar implementation, but changed it because the iteration with `equal_range` was somewhat obscure.

I added a description of the algorithm in the hopes to make it clearer.